### ISSUE TITLE
bio_write function appeared endless loop

### ISSUE THREAD

**Issue by ycf8788** [openssl/openssl#7181]
My server have appeared a question about endless loop by openssl 1.1.0f or 1.1.0i.
When the deadloop happened, I record the stack info , like this:

`19:40:03.441 d:\tddownload\code\openssl-1.1.0i_test\crypto\bio\bss_bio.c (325): bio_write`
`19:40:03.441 d:\tddownload\code\openssl-1.1.0i_test\crypto\bio\bio_lib.c (232): BIO_write`
`19:40:03.441 d:\tddownload\code\openssl-1.1.0i_test\ssl\record\rec_layer_s3.c (930): ssl3_write_pending`
`19:40:03.441 d:\tddownload\code\openssl-1.1.0i_test\ssl\record\rec_layer_s3.c (881): do_ssl3_write`

 I was write some test code into bio_write function in bss_bio.c

static int bio_write(BIO *bio, const char *buf, int num_)
{
    size_t num = num_;
    size_t rest;
    struct bio_bio_st *b;

    BIO_clear_retry_flags(bio);

    if (!bio->init || buf == NULL || num == 0)
        return 0;

    b = bio->ptr;
    assert(b != NULL);
    assert(b->peer != NULL);
    assert(b->buf != NULL);

    b->request = 0;
    if (b->closed) {
        /* we already closed */
        BIOerr(BIO_F_BIO_WRITE, BIO_R_BROKEN_PIPE);
        return -1;
    }

    assert(b->len <= b->size);

    if (b->len == b->size) {
        BIO_set_retry_write(bio); /* buffer is full */
        return -1;
    }

    /* we can write */
    if (num > b->size - b->len)
        num = b->size - b->len;

    /* now write "num" bytes */

    rest = num;
    size_t szInitRest = rest;

    assert(rest > 0);
    
    static int s_Temp1 = 0;
    static int s_Rest = 0;
    static int s_Chunk = 0;
	int nLoop = 0;
	
    do {                        /* one or two iterations */
        size_t write_offset;
        size_t chunk;

        assert(b->len + rest <= b->size);

        write_offset = b->offset + b->len;
        if (write_offset >= b->size)
            write_offset -= b->size;
        /* b->buf[write_offset] is the first byte we can write to. */

        if (write_offset + rest <= b->size)
            chunk = rest;
        else
            /* wrap around ring buffer */
            chunk = b->size - write_offset;

        memcpy(b->buf + write_offset, buf, chunk);

        b->len += chunk;

        assert(b->len <= b->size);

        if(rest < chunk)
        {
             s_Rest = rest;
             s_Chunk = chunk;
             printf("*****bio_write rest:%u, chunk:%u.*****\n", rest, chunk);
        }
		
       s_Temp1++;
       if(++nLoop >= 1000 || s_Temp1 % 5000 == 0)
       {
             printf("bio_write nLoop:%d, sTemp:%d, rest:%u, chunk:%u, s_rest:%u, s_chunk:%u, num_:%d.\n",
                   nLoop, s_Temp1, rest, chunk, s_Rest, s_Chunk, num_);
       }
			
        rest -= chunk;
        buf += chunk;
    }
    while (rest);

    return num;
}`

the console output these info:
bio_write nLoop:2227930, sTemp:2978994, rest:4294915158, chunk:0, s_res
t:43, s_chunk:52181, num_:43.
bio_write nLoop:2227931, sTemp:2978995, rest:4294915158, chunk:0, s_res
t:43, s_chunk:52181, num_:43.
bio_write nLoop:2227932, sTemp:2978996, rest:4294915158, chunk:0, s_res
t:43, s_chunk:52181, num_:43.

Do you have any suggestions or ideas, how to get around this?

Thanks!

**Comment by t8m** [comment]
I am sorry we neglected your question for so long time, is this still a problem?

**Comment by ljluestc** [comment]
```

do {
    size_t write_offset;
    size_t chunk;

    assert(b->len + rest <= b->size);

    write_offset = b->offset + b->len;
    if (write_offset >= b->size)
        write_offset -= b->size;

    if (write_offset + rest <= b->size)
        chunk = rest;
    else
        chunk = b->size - write_offset;

    if (chunk == 0) {
        /* Exit loop if chunk size is zero */
        break;
    }

    memcpy(b->buf + write_offset, buf, chunk);

    b->len += chunk;

    rest -= chunk;
    buf += chunk;
}
while (rest > 0);
```

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input
